### PR TITLE
improve api server boundary

### PR DESCRIPTION
**Pull request by jrhizor**
Improves the abstraction around the server creation so it's easier to mock various parts while building. It also separates shared OSS + Cloud configurations and OSS-specific configurations.

For Cloud, this will allows to implement a `ServerFactory` that configures the server that we will run with just `getServer(new CouldServerFactory()).start();` in a separate PR. We will also be able to run it in tests with mocked data on just a server running in a thread (which will allow us to do things like CORS header testing for the OSS project) by just running `new ServerFactory.Api().create(...).start()`.

### CODE REVIEW

File: airbyte-server/src/main/java/io/airbyte/server/ConfigurationApiBinder.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * SOFTWARE.
+ */
+
+package io.airbyte.server;
+
+import io.airbyte.server.apis.ConfigurationApi;
+import org.glassfish.hk2.utilities.binding.AbstractBinder;
+import org.glassfish.jersey.process.internal.RequestScoped;
+
+public class ConfigurationApiBinder extends AbstractBinder {
```
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/ConfigurationApiBinder.java:31]: interesting, learn about this today.

File: airbyte-server/src/main/java/io/airbyte/server/ServerFactory.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      ConfigurationApiFactory.setSynchronousSchedulerClient(cachingSchedulerClient);
+      ConfigurationApiFactory.setTemporalService(temporalService);
+      ConfigurationApiFactory.setConfigRepository(configRepository);
+      ConfigurationApiFactory.setJobPersistence(jobPersistence);
+      ConfigurationApiFactory.setConfigs(configs);
+      ConfigurationApiFactory.setArchiveTtlManager(new FileTtlManager(10, TimeUnit.MINUTES, 10));
+      ConfigurationApiFactory.setMdc(MDC.getCopyOfContextMap());
+
+      // server configurations
+      final Set<Class<?>> componentClasses = Set.of(ConfigurationApi.class);
```
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/ServerFactory.java:68]: I don't understand this. What is the difference between `componentClasses` and `component`?

File: airbyte-server/src/main/java/io/airbyte/server/ServerApp.java [hunk 3]
```diff
@@ -160,9 +119,9 @@ public void configure() {
             // https://stackoverflow.com/questions/35669774/jersey-custom-exception-mapper-for-invalid-json-string
             .register(JacksonJaxbJsonProvider.class);
 
-    // add filters
-    requestFilters.forEach(rc::register);
-    responseFilters.forEach(rc::register);
+    // inject custom server functionality
+    customComponentClasses.forEach(rc::register);
```
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/ServerApp.java:123]: what is the difference between `componentClass` and `component`? confused what to pass into what if we are starting up custom resources.

File: airbyte-server/src/main/java/io/airbyte/server/ServerApp.java [hunk 4]
```diff
@@ -254,15 +211,29 @@ public static void runServer(final Set<ContainerRequestFilter> requestFilters,
...
-      new ServerApp(configRepository, jobPersistence, configs, requestFilters, responseFilters).start();
+
+      final JobTracker jobTracker = new JobTracker(configRepository, jobPersistence);
+      final WorkflowServiceStubs temporalService = TemporalUtils.createTemporalService(configs.getTemporalHost());
+      final TemporalClient temporalClient = TemporalClient.production(configs.getTemporalHost(), configs.getWorkspaceRoot());
+      final SchedulerJobClient schedulerJobClient = new DefaultSchedulerJobClient(jobPersistence, new DefaultJobCreator(jobPersistence));
+      final DefaultSynchronousSchedulerClient syncSchedulerClient = new DefaultSynchronousSchedulerClient(temporalClient, jobTracker);
+      final SpecCachingSynchronousSchedulerClient cachingSchedulerClient = new SpecCachingSynchronousSchedulerClient(syncSchedulerClient);
+
+      return apiFactory.create(
```
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/ServerApp.java:222]: What about simplifying this constructor by passing in only `configRepository`, `jobPersistence` and `configs`? L214 - 222 can be pushed into the  `ServerFactory`.

File: airbyte-server/src/main/java/io/airbyte/server/ServerApp.java [hunk 5]
```diff
@@ -254,15 +211,29 @@ public static void runServer(final Set<ContainerRequestFilter> requestFilters,
...
     } else {
       LOGGER.info("Start serving version mismatch errors. Automatic migration either failed or didn't run");
-      new VersionMismatchServer(airbyteVersion, airbyteDatabaseVersion.get(), PORT).start();
+      return new VersionMismatchServer(airbyteVersion, airbyteDatabaseVersion.get(), PORT);
     }
   }
 
   public static void main(String[] args) throws Exception {
-    runServer(Collections.emptySet(), Set.of(new CorsFilter()));
+    getServer(new ServerFactory.Api()).start();
```
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/ServerApp.java:236]: should we split up `getServer`? 

reads like two separate methods now:
1) `verifyAndMigrate`
2) `startServer`

thinking out loud.

**Review by davinchia — APPROVED** [review]
Nice! Took me some time to grok, but definitely think this is better.

Some question for my understand and 2 suggestions for readability. None blocking.

File: airbyte-server/src/main/java/io/airbyte/server/ServerFactory.java (hunk 2 above)
Comment (jrhizor) [reply airbyte-server/src/main/java/io/airbyte/server/ServerFactory.java:68]: One provides specific instances of componenets, the other provides classes that can be instantiated on demand by the server.

File: airbyte-server/src/main/java/io/airbyte/server/ServerApp.java (hunk 3 above)
Comment (jrhizor) [reply airbyte-server/src/main/java/io/airbyte/server/ServerApp.java:123]: See above. It depends on if it's a shared class or spun up for requests.

File: airbyte-server/src/main/java/io/airbyte/server/ServerApp.java (hunk 4 above)
Comment (jrhizor) [reply airbyte-server/src/main/java/io/airbyte/server/ServerApp.java:222]: You can't reliably create these clients based off of mocks() if they have validations internally, which makes it much harder to inject. I'm planning on using this functionality in tests soon.

File: airbyte-server/src/main/java/io/airbyte/server/ServerApp.java (hunk 5 above)
Comment (jrhizor) [reply airbyte-server/src/main/java/io/airbyte/server/ServerApp.java:236]: Agreed, but there is sequential object creation throughout `verifyAndMigrate` where we'd need to return a context (?) that contains the various objects we're instantiating, and consuming that in `startServer`?

Comment (jrhizor) [reply airbyte-server/src/main/java/io/airbyte/server/ServerApp.java:236]: Agreed, but there is sequential object creation throughout `verifyAndMigrate` where we'd need to return a context (?) that contains the various objects we're instantiating, and consuming that in `startServer`?

**Comment by jrhizor** [comment]
Going to merge for now. Let me know if you still want me to revisit your comments @davinchia

File: airbyte-server/src/main/java/io/airbyte/server/ServerApp.java (hunk 4 above)
Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/ServerApp.java:222]: Fair. I think I was thinking of two constructors,  a slimmer down one for production and an exposed one for testing. We can bolt this on later.

**Comment by davinchia** [comment]
@jrhizor my comments are cosmetic, we good.

File: airbyte-server/src/main/java/io/airbyte/server/ServerFactory.java (hunk 2 above)
Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/ServerFactory.java:68]: I feel like we should drop a comment here

**Review by cgardens — COMMENTED** [review]
cool

**Pull request merged**